cou/steps/plan.py

```diff
@@ -653,22 +658,54 @@ def _create_upgrade_group(
     :return: Upgrade plan of an upgrade group.
     :rtype: UpgradePlan
     """
-    plan_warnings = PlanWarnings()
     group_upgrade_plan = UpgradePlan(description)
 
     for app in apps:
-        try:
-            app_upgrade_plan = app.generate_upgrade_plan(target, force)
+        if app_upgrade_plan := _generate_instance_plan(app, target, force):
             group_upgrade_plan.add_step(app_upgrade_plan)
-        except HaltUpgradePlanGeneration as exc:
-            # we do not care if applications halt the upgrade plan generation
-            # for some known reason.
-            logger.debug("'%s' halted the upgrade planning generation: %s", app.name, exc)
-        except COUException as exc:
-            logger.debug("Cannot generate plan for '%s'\n\t%s", app.name, exc)
-            plan_warnings.add_warning(f"Cannot generate plan for '{app.name}'\n\t{exc}")
-        except Exception as exc:
-            logger.error("Cannot generate upgrade plan for '%s': %s", app.name, exc)
-            raise
 
     return group_upgrade_plan
+
+
+def _generate_instance_plan(
+    instance: Union[HypervisorUpgradePlanner, OpenStackApplication],
+    target: OpenStackRelease,
+    force: bool,
+) -> Optional[UpgradePlan]:
+    """Generate upgrade plan for an instance and handle exceptions.
+
+    COUExceptions (except HaltUpgradePlanGeneration) raised by application will be stored
+    in the PlanWarnings object.
+
+    :param instance: An OpenStackApplication or HypervisorUpgradePlanner instance to generate
+                     plan for.
+    :type instance: Union[HypervisorUpgradePlanner, OpenStackApplication]
+    :param target: Target OpenStack release.
+    :type target: OpenStackRelease
+    :param force: Whether the plan generation should be forced
+    :type force: bool
+    :raises Exception: When cannot generate upgrade plan.
+    :return: Upgrade plan of an instance.
+    :rtype: Optional[UpgradePlan]:
+    """
+    instance_id = (
+        instance.name
+        if isinstance(instance, OpenStackApplication)
+        else "Hypervisors Groups:" + ", ".join(instance.get_azs(target).keys())
+    )
+
+    try:
+        instance_upgrade_plan = instance.generate_upgrade_plan(target, force)
+        return instance_upgrade_plan
+    except HaltUpgradePlanGeneration as exc:
+        # we do not care if applications halt the upgrade plan generation
+        # for some known reason.
+        logger.debug("'%s' halted the upgrade planning generation: %s", instance_id, exc)
+    except COUException as exc:
+        logger.debug("Cannot generate plan for '%s'\n\t%s", instance_id, exc)
+        PlanWarnings().add_warning(f"Cannot generate plan for '{instance_id}'\n\t{exc}")
```

Comment:
I thought we will return here the "place holder" plan. Did we decide to change it?
not-blocking: It would not be easier to define debug log message in `add_warning`?

Reply:
Actually I just read through the code for step class: if the step and none of its sub steps contain coroutine, then the step will be considered as `false` and therefore will not show up in the plan. So creating a `place-holder` plan would not work as we want. So I'm in favor of keeping it the way it is now.

Reply:
We discuss this on sync meeting and I think it's ok. resolving. Thanks